src/katsdpmetawriter/__init__.py

```diff
@@ -170,52 +173,57 @@ def get_s3_connection(boto_dict, fail_on_boto=False):
     s3_conn : S3Connection
         A connection to the s3 endpoint. None if a connection error occurred.
     """
-    s3_conn = boto.connect_s3(**boto_dict)
+    session = aiobotocore.session.get_session()
     try:
         # reliable way to test connection and access keys
-        s3_conn.get_canonical_user_id()
-        return s3_conn
+        async with session.create_client('s3', **botocore_dict) as s3_conn:
+            await s3_conn.list_buckets()
+            yield s3_conn
+            return
     except socket.error as e:
         logger.error(
-            "Failed to connect to S3 host %s:%s. Please check network and host address. (%s)",
-            boto_dict['host'], boto_dict['port'], e)
-    except boto.exception.S3ResponseError as e:
-        if e.error_code == 'InvalidAccessKeyId':
+            "Failed to connect to S3 host %s. Please check network and host address. (%s)",
+            botocore_dict['endpoint_url'], e)
+    except botocore.exceptions.ClientError as e:
+        error_code = e.response.get('Error', {}).get('Code')
+        if error_code == 'InvalidAccessKeyId':
             logger.error(
                 "Supplied access key %s is not a valid S3 user.",
-                boto_dict['aws_access_key_id'])
-        if e.error_code == 'SignatureDoesNotMatch':
+                botocore_dict['aws_access_key_id'])
+        elif error_code == 'SignatureDoesNotMatch':
             logger.error("Supplied secret key is not valid for specified user.")
-        if e.status == 403 or e.status == 409:
+        elif error_code == 'AccessDenied':
             logger.error(
                 "Supplied access key (%s) has no permissions on this server.",
-                boto_dict['aws_access_key_id'])
+                botocore_dict['aws_access_key_id'])
+        else:
+            logger.error(e)
         if fail_on_boto:
             raise
-    return None
+    yield None
```

Comment:
What does the second `yield` do? Is that to cater for the exceptions?

Reply:
Yes, if `fail_on_boto` is false and there is an exception it "returns" None (not a design I liked, but we can blame Simon). But for `contextlib.contextmanager`, one uses yield to "return" a value to the `foo` is `with get_s3_connection(...) as foo`.

Reply:
Thanks, makes sense.